jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/CodeWhispererTelemetryServiceTest.kt

```diff
@@ -425,6 +445,146 @@ class CodeWhispererTelemetryServiceTest {
+        )
+
+        val completionType = if (recommendationContext.details.any {
+                it.completionType == CodewhispererCompletionType.Block
+            }
+        ) {
+            CodewhispererCompletionType.Block
+        } else CodewhispererCompletionType.Line
+
+        verify(mockClient).putUserTriggerDecisionTelemetry(
+            eq(requestContext),
+            eq(responseContext),
+            eq(completionType),
+            eq(suggestionState),
+            eq(suggestionReferenceCount),
+            eq(lineCount),
+        )
+    }
+
+    @Test
+    fun `should not invoke putTelemetryEvent if opt out telemetry`() {
+        val project = projectRule.project
+
+        val mockClient = mock<CodeWhispererClientAdaptor>()
+        project.replaceService(
+            CodeWhispererClientAdaptor::class.java,
+            mockClient,
+            disposableRule.disposable
+        )
+
+        val mockStatesManager = mock<CodeWhispererExplorerActionManager>()
+        ApplicationManager.getApplication().replaceService(
+            CodeWhispererExplorerActionManager::class.java,
+            mockStatesManager,
+            disposableRule.disposable
+        )
+
+        fun configureMock(isProTier: Boolean, isTelemetryEnabled: Boolean) {
+            if (isProTier) {
+                whenever(mockStatesManager.checkActiveCodeWhispererConnectionType(any())).doReturn(CodeWhispererLoginType.SSO)
+            } else {
+                whenever(mockStatesManager.checkActiveCodeWhispererConnectionType(any())).doReturn(CodeWhispererLoginType.Sono)
+            }
+
+            AwsSettings.getInstance().isTelemetryEnabled = isTelemetryEnabled
+        }
+
+        // case 1
+        configureMock(isProTier = true, isTelemetryEnabled = false)
+        sut.sendUserTriggerDecisionEvent(
+            aRequestContext(project),
+            aResponseContext(),
+            aRecommendationContext(),
+            CodewhispererSuggestionState.Accept,
+            Duration.ofSeconds(
+                1
+            ),
+            1,
+            50
+        )
+
+        verifyNoInteractions(mockClient)
+
+        // case 2
+        configureMock(isProTier = false, isTelemetryEnabled = true)
+        sut.sendUserTriggerDecisionEvent(
+            aRequestContext(project),
+            aResponseContext(),
+            aRecommendationContext(),
+            CodewhispererSuggestionState.Accept,
+            Duration.ofSeconds(
+                1
+            ),
+            1,
+            50,
+        )
+
+        verifyNoInteractions(mockClient)
+
+        // case 3
+        configureMock(isProTier = false, isTelemetryEnabled = false)
+        sut.sendUserTriggerDecisionEvent(
+            aRequestContext(project),
+            aResponseContext(),
+            aRecommendationContext(),
+            CodewhispererSuggestionState.Accept,
+            Duration.ofSeconds(
+                1
+            ),
+            1,
+            50
+        )
```

Comment:
shall we have a `verifyNoInteractions(mockClient)` here?

Reply:
I think so, let me add it.